trin-core/src/utp/stream.rs

```diff
@@ -766,35 +844,337 @@ impl UtpSocket {
+        assert!(response.is_some());
+        let response = response.unwrap();
+        assert_eq!(response.get_type(), PacketType::State);
+
+        let old_packet = packet;
+        let old_response = response;
+
+        // Now, send a keepalive packet
+        let mut packet = Packet::new();
+        packet.set_wnd_size(BUF_SIZE as u32);
+        packet.set_type(PacketType::State);
+        packet.set_connection_id(initial_connection_id);
+        packet.set_seq_nr(old_packet.seq_nr() + 1);
+        packet.set_ack_nr(old_response.seq_nr());
+
+        let response = socket.handle_packet(&packet, client_enr.clone());
+        assert!(response.is_ok());
+        let response = response.unwrap();
+        assert!(response.is_none());
+
+        // Send a second keepalive packet, identical to the previous one
+        let response = socket.handle_packet(&packet, client_enr.clone());
+        assert!(response.is_ok());
+        let response = response.unwrap();
+        assert!(response.is_none());
+
+        // Mark socket as closed
+        socket.state = SocketState::Closed;
+    }
+
+    #[tokio::test]
+    async fn test_response_to_wrong_connection_id() {
+        // Boilerplate test setup
+        let initial_connection_id: u16 = rand::random();
+        let client_enr = generate_random_enr();
+        let server_enr = generate_random_enr();
+        let server_config = create_portal_config();
+        let mut server_discv5 = Discovery::new(server_config).unwrap();
+        server_discv5.start().await.unwrap();
+        let mut socket = UtpSocket::new(Arc::new(server_discv5), server_enr.clone());
+
+        // Establish connection
+        let mut packet = Packet::new();
+        packet.set_wnd_size(BUF_SIZE as u32);
+        packet.set_type(PacketType::Syn);
+        packet.set_connection_id(initial_connection_id);
+
+        let response = socket.handle_packet(&packet, client_enr.clone());
+        assert!(response.is_ok());
+        let response = response.unwrap();
+        assert!(response.is_some());
+        assert_eq!(response.unwrap().get_type(), PacketType::State);
+
+        // Now, disrupt connection with a packet with an incorrect connection id
+        let new_connection_id = initial_connection_id.wrapping_mul(2);
+
+        let mut packet = Packet::new();
+        packet.set_wnd_size(BUF_SIZE as u32);
+        packet.set_type(PacketType::State);
+        packet.set_connection_id(new_connection_id);
+
+        let response = socket.handle_packet(&packet, client_enr);
+        assert!(response.is_ok());
+        let response = response.unwrap();
+        assert!(response.is_some());
+
+        let response = response.unwrap();
+        assert_eq!(response.get_type(), PacketType::Reset);
+        assert_eq!(response.ack_nr(), packet.seq_nr());
+
+        // Mark socket as closed
+        socket.state = SocketState::Closed;
+    }
+
+    #[tokio::test]
+    async fn test_unordered_packets() {
+        // Boilerplate test setup
```

Comment:
Any reason not to pull out this setup into a single helper function? It seems to be used across all the tests here